File: src/components/conversation/journal.js
/* What a driven session's journal is drawn as, and whether a turn is in flight.
   Both are whole rules and neither is in `ConversationView.vue`, for the reason
   this whole family of modules is outside the component that draws it: a `.vue`
   file is the one thing no runner in this repository can reach, so a rule left
   inside one is a rule nothing checks.

   Pure: no Vue, no DOM, no store. The events are the worker's own
   (`src-tauri/src/session/model.rs`), unchanged, and what comes back is a list
   of rows a template can `v-for` over. */

/* The wire's names, written out once for the whole front end.

   `session::model::Event` flattens its kind's fields in beside a kebab-case
   `kind` and renames none of them — `tokens_in`, `tokens_out`, `cost_usd` —
   which is deliberate on that side and carries a comment saying so. Every
   numeric prop of `TurnResult` has a default, so an event handed over raw
   (`v-bind="event"`, or `:tokens-in="event.tokens_in"` read off a store with no
   translation) draws `0 in · 0 out · 0 ms` silently: no Vue warning, nothing on
   screen to say the numbers are not the session's, and no component test in
   this project to catch it. That is why the translation is here, where a test
   can read it, and why there is exactly one copy of it.

   **An event kind this does not know produces no row.** The chain is closed and
   has no fallback, which is `EventKind`'s own rule one layer up — a missing row
   costs a person nothing the harness's own logs do not still hold, while a wall
   of raw protocol costs them the panel. The two permission kinds are drawn from
   the session's open question, at the foot of the panel where it is answered,
   rather than twice.

   A tool call and its result are **one row**, folded by the id they share:
   `tool-result` carries no name and no detail of its own, so a row of its own
   would split one thing the agent did across two. `result: null` is the running
   state and is what `ToolCall` is written against — it draws the live glyph
   rather than guessing at an outcome. A result whose call is not in this
   journal is dropped: that is a journal the worker has trimmed, and there is
   nothing to put in the row.

   **`turn-start` no longer produces nothing.** It used to, on the reading that
   the message under it already says a turn began; what it opens now is the
   strip `markup-contract.md` section 6 calls the agent's activity — one
   element, four moments (`streaming` joined `waiting`, `done` and `failed`
   at smetana-6we6), folded here exactly as a tool call and its result are.
   `result` closes it `done`, with the wire's own `tokens_in`, `tokens_out`,
   `cost_usd` and `ms`, the same four this file has always translated. `error`
   closes it `failed` **only while a turn is actually open** — `session::model`
   also uses `EventKind::Error` for a message that never reached a session with
   nothing open at all (`Request::Send` against a dead child), and that one is
   not a turn ending, it is the worker saying where the words went instead; it
   keeps the old bare `error` row a turn's fold has no share in.

   **A turn still open when the batch ends is not always `waiting`.** The
   commoner of the two is a live one — `waiting`, carrying the `turn-start`'s
   own `at` rather than a duration, since nothing pure can compute how long
   "now" is, which is `TurnResult.vue`'s own clock to tick. But `Chunk::Eof`
   (`session::service`, an agent process ending on its own) sets the session's
   own state `failed` while appending **no** `Error` at all — closing a shell
   window, and `Stop` itself, both end there, since `ClaudeDriver::interrupt`
   answers `None` and `Request::Stop` reaches for `start_kill()` outright. A
   turn read against the events alone would stay `waiting` forever in that
   case, its clock still climbing next to a header that already reads
   `failed` and a composer already reading Send again — which is why `state`
   is the second argument here.

   **The check is narrow on purpose: `state === 'failed' || state === 'exited'`,
   never the wider `!isBusy(state)`.** `state_of` (`session::model`) is the
   authority on what those words can mean: with the child dead it answers
   `Failed` for an open turn and `Exited` for a closed one, so `exited` with a
   turn still open is a state that cannot truthfully arise and is named here
   only so a value this file has misread can never reach it either. `ready` and
   `starting` are the ordinary skew of two events delivered as two IPC
   messages — `append()` ships `session:events` and only then calls
   `refresh_state`, which ships `session:state`
   (`service.rs`) — so a fresh `turn-start`/`user-message` can sit for a
   reactive flush against a `state` that has not moved off `ready` yet. Read
   against `!isBusy`, that skew closed the strip `failed`, wordless and
   saturated, under the very message that just opened the turn, on every
   single send. Read against this narrower pair it ticks `waiting` instead —
   the honest answer for a state this fold cannot yet see has ended, which is
   the same conservative reading `BUSY`'s own header gives a word it has never
   heard of.

   **The closed turn still gets words of its own.** `text` is not the empty
   string here any more: nobody said anything, but the panel knows one true
   sentence — the session ended while this turn was still open — and prints
   that rather than a blank strip a person would read as broken rather than
   stopped. This is the strip's only reader of `state` — every other row is
   the events alone.

   **`streaming` is the fourth moment (smetana-6we6), and it is derived from
   the same events rather than carried on the wire as a state of its own.**
   `text-delta` is `session::model::EventKind::TextDelta` — one incremental
   piece of the reply, never accumulated on the Rust side — and this fold is
   what stitches a run of them into one growing `agent` row, `streamingRow`
   below. The stitched row's `key` is the *first* delta's own `seq` and never
   changes again: Vue's `:key` is what decides whether a row is the same DOM
   node or a new one, and a key that moved when the reply finished would be
   the reflow the acceptance criteria explicitly rule out. Closing a stream is
   the existing whole `text` event doing exactly what it always did — this
   fold does not wait for a second "done streaming" signal, because there
   isn't one and none is needed: the wire's `Text` already arrives once a
   content block completes (`claude_driver.rs`'s own header), so it is both
   the reply's authoritative content and its closing bell. On close the
   stitched text is **replaced wholesale** by the whole event's own copy
   rather than trusted to equal the concatenation — cheap, and immune to the
   two ever silently drifting.

   A row still being streamed carries `streaming: true`, which is the one
   thing that tells `AgentMessage`/`Markdown` to draw the live edge
   (`span[data-edge]`, contract section 6) after its last character — see
   `Markdown.vue`'s own header for where. Five places close a row still
   streaming, setting `streaming: false` without touching its `text` again:
   the ordinary close (`text` above), `result`, `error`, the trailing
   `CHILD_GONE` branch, and the trailing `else if (streamingRow)` beside it —
   because a caret pinned to a reply nothing is still writing is the exact
   scrap the acceptance criteria refuse to leave behind, and every one of the
   five is unconditional on whether this fold can still see the turn that
   opened the stream. That last qualification is load-bearing rather than
   defensive phrasing: `openAt` tracks the *turn* (set at `turn-start`,
   cleared at `result` or at a turn-closing `error`), and `streamingRow`
   tracks the *stream*, and the two can come apart once a journal is long
   enough to trim the `turn-start` off its front while later `text-delta`s
   survive — a trim drops from the front regardless of which kind of event it
   meets, and a reply's own deltas no longer accumulate past its own close
   (`journal.rs`'s own `Journal::append`), so this is reachable rather than
   ordinary: it takes one reply's still-open stream alone reaching `BUDGET`
   deltas to evict its own `turn-start`. `error`'s clear is therefore
   outside its own `openAt != null` branch, and the trailing `else if` exists
   for the same reason applied to the tail of the fold: a `streamingRow` still
   open when the loop ends with no `openAt` at all has nothing above it left
   to close it otherwise. In the protocol this wire actually speaks a content
   block is always closed by its own whole event before the next one opens,
   so in the ordinary case `streamingRow` is already `null` by the time any of
   the five is reached — the explicit closes are what stop a truly abandoned
   stream (the child dying mid-chunk, or its turn-start trimmed out from under
   it) from leaving a pulsing mark on screen forever, not a shape this file
   expects to hit on the every-day path.

   **A re-entry never finds a stitched row waiting to be resumed, and that
   rests on two separate guarantees rather than one.** Re-attaching to a
   session whose worker is still alive asks for a fresh snapshot
   (`session::journal::Journal::snapshot`), and what that snapshot holds
   depends on whether the stream it would be replaying is still open. A reply
   genuinely still arriving is handed over exactly as its own deltas stand
   right now, so it draws here exactly as it would have live — correct, since
   it is live. **A reply that has already closed is not handed over as the
   run of deltas a live listener actually watched arrive, one at a time** —
   `Journal::append` drops a stream's deltas the instant its closing `text`
   lands beside them, so a re-attacher's snapshot already holds only the
   consolidated event, the same way a fresh session opens on one. That
   collapse is safe *because* this fold treats the two shapes identically:
   a `text` with no `streamingRow` open behind it and a `text` closing one
   both end in the same finished row, so a re-attacher seeing fewer raw events
   than a live listener once did draws the same thing regardless. A session
   read back from Claude Code's own transcript (`session::history`, after a
   restart) never carries a `text-delta` at all in the first place: that
   harness's `.jsonl` holds only the consolidated `assistant` records this
   fold already turns into `text`, never the raw `stream_event` lines partial
   messages ride on — checked against the installed CLI rather than assumed,
   and `history::is_past` refuses the event kind a second time regardless. So
   there is no event sequence this fold could ever see that ends on an
   unclosed `streamingRow`, other than the live, still-open turn the trailing
   branch below already handles on purpose.

   **The mirror of that Rust-side collapse does not belong on this side, and
   that is worth writing down before somebody reaches for it.**
   `stores/conversation.js`'s `held.events` is push-only for the life of an
   attachment — it never had a budget and this task did not give it one, so a
   window open long enough keeps every delta it was ever sent. That is not
   this fold's problem to solve, but the obvious fix looks like one:
   dropping a held `text-delta` the moment its `text` arrives would make this
   very fold take the *plain* `text` branch instead of the closing one, which
   keys the new row on the `text` event's own `seq` rather than the first
   delta's. That is the reflow the acceptance criteria protect against,
   reintroduced by the fix meant to save memory — a row remounting at exactly
   the moment it finishes. Any future trim of `held.events` has to preserve
   the *key* a stream opened under, not merely the words.

   **The elapsed figure is the gap to the *last event this batch still
   holds*, not to the moment the child actually died** — `journal.js` has no
   later timestamp to read, and a batch holding only the `turn-start` reads
   `0s` rather than nothing. That under-reports whenever the agent's last
   tool call landed well before the process actually ended, sometimes by
   minutes; it never over-reports, and unlike anything measured off
   `Date.now()` it does not change if the panel is closed and reopened, which
   matters for a number sitting in a journal rather than ticking on screen.

   `Error` carries no duration of its own — `session::model::EventKind::Error`
   is `{ text }`, full stop — so a `failed` turn's elapsed time is the one this
   file *can* compute, the gap between the `turn-start` that opened it and the
   `error` that closed it. `elapsedSince` is that same gap asked of anything
   produced while a turn is open, which is also what gives `Reasoning`'s own
   `<summary><time>` a number to show: `EventKind::Reasoning` carries no
   duration either, and "how long the agent had been going when it said this"
   is the honest reading available from two timestamps the wire already sends.
   `null` where no turn is open, which the component reads as nothing to print
   rather than a guess at zero.

   `key` is the event's own `seq`, which the worker mints from one counter per
   session and never reuses — except the trailing `waiting` or crash-closed
   `failed`, both keyed by the `turn-start` that opened the turn they are
   still describing, since neither has a closing event of its own to mint a
   key under. */

/* Whether the agent is working, which is what turns the composer's one button
   into Stop — moved above `journalRows` because its trailing branch is now a
   second reader of it.

   **A closed list of `session::model::SessionState`'s wire words, and the
   sharpest reason anything here is outside the component.** Rename `Running` on
   the Rust side and a `.vue` file holding this list would stop showing the Stop
   button for the whole of a turn — a person could fire messages into a working
   agent with no way to stop it, and every gate in this repository would stay
   green. Here, the list is one grep and one test away from whoever renames it.

   `needs-you` counts as busy: the turn is open and what to do about it is the
   permission card above the field, not another message. `starting`, `ready`,
   `exited` and `failed` are the four that are not, and a word this front end
   has never heard of is not either — the honest reading of "we do not know that
   the agent is working" is to leave the person able to type.

   **`starting` is on that side, and the reason is worth keeping written down.**
   A driven session is started with `--input-format stream-json`, and a harness
   parked on its own stdin says nothing whatever: the journal stays empty, which
   is what `state_of` in `session::model` calls `starting`. Counted as a turn in
   flight, that state closed the only road out of itself — it ends at the first
   event, the first event is the `TurnStart` the worker appends when a message
   is sent, and the button that sends one had become Stop. A new agent sat at an
   empty panel with a composer that refused every key, for good. */
const BUSY = ['running', 'needs-you']

export const isBusy = (state) => BUSY.includes(state)

/* Closes an open turn even though nothing in its own events did — the two
   states `state_of` can only reach with the child already dead. Not
   `!isBusy(state)`: `ready` and `starting` are the ordinary skew between
   `session:events` and the `session:state` that follows it, and reading that
   skew as a death closed the strip `failed` under the very message that had
   just opened the turn. See the fold's own header for the fuller reasoning. */
const CHILD_GONE = ['failed', 'exited']

/* The one sentence the panel can truthfully say about a turn the events never
   closed: not what happened, since the worker gave no words for that, but
   that the session ended while this turn was still open. A blank strip reads
   as broken rather than stopped, which is the wordless version this
   replaces. */
const TURN_ENDED = 'The session ended while this turn was still open.'

export function journalRows(events = [], state) {
  const rows = []
  const calls = new Map()
  /* The turn open right now, if any — never more than one, since a person
     cannot send a second message while the composer's one button reads Stop.
     Cleared the moment `result` or `error` closes it, and read once more after
     the loop if nothing did. */
  let openAt = null
  let openSeq = null
  /* The `agent` row currently taking `text-delta`s, or `null` between
     streams — see this file's own header. Cleared (with `streaming` set
     `false` first) by the same four places that close a turn, and read once
     more after the loop for the same reason `openAt` is. */
  let streamingRow = null
  /* `null` rather than `NaN` where either stamp is missing — the wire always
     sends one, but a `NaN` would pass `TurnResult`'s `type: Number` check and
     draw `0s`, indistinguishable from a turn that genuinely took none. */
  const elapsedSince = (at) => (openAt == null || at == null ? null : Date.parse(at) - Date.parse(openAt))

  for (const event of events) {
    if (event.kind === 'turn-start') {
      openAt = event.at
      openSeq = event.seq
    } else if (event.kind === 'user-message') {
      rows.push({
        key: event.seq,
        kind: 'user',
        text: event.text,
        attachments: event.attachments ?? []
      })
    } else if (event.kind === 'text-delta') {
      if (streamingRow) {
        streamingRow.text += event.text
      } else {
        streamingRow = { key: event.seq, kind: 'agent', text: event.text, streaming: true }
        rows.push(streamingRow)
      }
    } else if (event.kind === 'text') {
      /* The stream's own close, not a second row: the stitched text is
         replaced wholesale by the wire's own authoritative copy rather than
         trusted to equal the concatenation, and the row keeps the key its
         first delta minted so nothing here reflows when a reply finishes. A
         harness that never streams (today: any history replay, and Codex's
         own PTY road never reaches this fold at all) still takes the plain
         branch exactly as before this task. */
      if (streamingRow) {
        streamingRow.text = event.text
        streamingRow.streaming = false
        streamingRow = null
      } else {
        rows.push({ key: event.seq, kind: 'agent', text: event.text })
      }
    } else if (event.kind === 'reasoning') {
      rows.push({ key: event.seq, kind: 'reasoning', text: event.text, ms: elapsedSince(event.at) })
    } else if (event.kind === 'tool-use') {
      const row = {
        key: event.seq,
        kind: 'tool',
        name: event.name,
        detail: event.detail,
        result: null
      }
      calls.set(event.id, row)
      rows.push(row)
    } else if (event.kind === 'tool-result') {
      const row = calls.get(event.id)
      if (row) row.result = { ok: event.ok, summary: event.summary }
    } else if (event.kind === 'result') {
      // A defensive close, not the ordinary road: this wire's protocol closes
      // a text block with its own `text` before a turn's `result` can follow,
      // so `streamingRow` is already `null` here in the every-day case. What
      // this guards is the shape that protocol does not promise — a result
      // arriving with a stream never properly closed — and it takes the same
      // "keep the words, drop the caret" reading as every other close.
      if (streamingRow) {
        streamingRow.streaming = false
        streamingRow = null
      }
      rows.push({
        key: event.seq,
        kind: 'activity',
        state: 'done',
        tokensIn: event.tokens_in,
        tokensOut: event.tokens_out,
        costUsd: event.cost_usd ?? null,
        ms: event.ms
      })
      openAt = null
      openSeq = null
    } else if (event.kind === 'error') {
      // The caret close is deliberately outside the `openAt != null` branch
      // below: `openAt` tracks the *turn*, not the stream, and the two can
      // come apart once a journal is long enough to trim a `turn-start` away
      // while a `text-delta` appended after it survives. A `streamingRow`
      // orphaned that way must still lose its caret on any event that says
      // the words stopped, whether or not this fold can still see the turn
      // that opened them.
      if (streamingRow) {
        streamingRow.streaming = false
        streamingRow = null
      }
      if (openAt != null) {
        rows.push({ key: event.seq, kind: 'activity', state: 'failed', text: event.text, ms: elapsedSince(event.at) })
        openAt = null
        openSeq = null
      } else {
        rows.push({ key: event.seq, kind: 'error', text: event.text })
      }
    }
  }

  if (openAt != null) {
    if (CHILD_GONE.includes(state)) {
      /* The events said nothing closed this turn, but the session's own state
         already has — `Chunk::Eof` with nothing journalled, which is `Stop`,
         a shell window closing, or any other end that rang no `Error`. `ms` is
         the gap to the last event this batch actually holds, in place of the
         one a real `error` would have closed on. A stream still open at this
         point is the same abandoned-reply shape: the words stay, the caret
         goes, for the reason the two branches above already give. */
      if (streamingRow) streamingRow.streaming = false
      rows.push({
        key: openSeq,
        kind: 'activity',
        state: 'failed',
        text: TURN_ENDED,
        ms: elapsedSince(events[events.length - 1].at)
      })
    } else if (streamingRow) {
      /* The fourth moment: the same strip `waiting` stands in, at the same
         `startedAt` — a stream is still the one open turn, and switching the
         clock to a second one at the first delta would tick from zero right
         next to a reply already several seconds in. */
      rows.push({ key: openSeq, kind: 'activity', state: 'streaming', startedAt: openAt })
    } else {
      rows.push({ key: openSeq, kind: 'activity', state: 'waiting', startedAt: openAt })
    }
  } else if (streamingRow) {
    // `openAt` tracks the turn, not the stream, and a long enough reply can
    // trim its own `turn-start` off the front of the journal while its own
    // `text-delta`s, appended later, survive — `openAt` reads `null` here
    // with `streamingRow` still open. There is no turn left to draw an
    // activity strip for, but the row already in `rows` still carries a
    // caret nothing above this point would otherwise ever clear.
    streamingRow.streaming = false
  }

  return rows
}

File: src-tauri/src/session/journal.rs
//! The ordered event log of one session. This is what `terminal::ring` is for a
//! terminal — the thing a window that has just opened is caught up from — with
//! events in place of bytes.

use std::collections::HashSet;

use super::model::{Event, EventKind};

/// How many events a session keeps. A long night of tool calls is thousands of
/// them and each is small; the ceiling is here so that a session left running
/// for a week cannot grow without bound.
///
/// **That "each is small" is no longer true of every event, and `append`'s own
/// `TextDelta` collapse below is what keeps it true in practice.** One
/// substantial streamed reply is several hundred `TextDelta`s where it used to
/// be a single `Text` (smetana-6we6); left uncollapsed, this budget would be
/// spent by five to fifteen ordinary turns rather than the "long night" or
/// "week" this comment has always promised — a promise that was never
/// re-measured against the granularity the wire actually carries once it
/// changed. It holds again only because a closed reply's deltas do not stay in
/// the journal once it closes.
pub const BUDGET: usize = 4000;

pub struct Journal {
    events: Vec<Event>,
    next: u64,
    /// The highest `seq` trimming has actually taken. Recorded rather than
    /// inferred from the oldest event still held, because trimming pins an
    /// unanswered `Permission` at the front and drops what follows it: the
    /// journal is then not a contiguous suffix, and its first event's `seq`
    /// says nothing about where the gap ends. Inferring it meant `since` could
    /// never report a gap again once a question was pinned.
    trimmed_through: u64,
}

impl Journal {
    pub fn new() -> Self {
        // Numbering opens at 1 so that 0 can mean "nothing seen yet" — a front
        // end that has never attached asks from 0 and must be given everything.
        Self { events: Vec::new(), next: 1, trimmed_through: 0 }
    }

    pub fn append(&mut self, kind: EventKind, at: String) -> Event {
        let event = Event { seq: self.next, at, kind };
        self.next += 1;
        self.events.push(event.clone());
        // The `TextDelta`s a closing `Text` supersedes do not go on living in
        // the journal beside it. Without this, one substantial streamed reply
        // is several hundred events where it used to be one, and `BUDGET`
        // arrives in five to fifteen turns rather than the week this file's
        // own comment on it promises — see that comment for the reasoning in
        // full.
        //
        // **`trimmed_through` is deliberately untouched.** That field is
        // `since`'s floor: below it, a cursor is told to re-snapshot rather
        // than trust a vec with a hole in it. Moving it here would claim the
        // same thing `trim` claims — "everything below this line is gone for
        // a reason a caller must recover from" — which is false of a delta a
        // live client already received in full. `since(seq)` only ever
        // filters by `event.seq > seq`; it does not ask whether a particular
        // `seq` is present, so a delta's absence from `self.events` is
        // invisible to a cursor holder who long since moved past it, and a
        // fresh attacher's `snapshot()` simply never sees deltas a reply has
        // already closed over.
        //
        // A stream can only ever have one reply's worth of deltas held at
        // once, by construction: the very last time a `Text` was appended, it
        // ran this same retain and cleared every delta that came before it.
        // So this always removes exactly the run just superseded, never more.
        if matches!(&event.kind, EventKind::Text { .. }) {
            self.events.retain(|held| !matches!(held.kind, EventKind::TextDelta { .. }));
        }
        self.trim();
        event
    }

    /// Everything after `seq`, or `None` when `seq` is older than what is still
    /// held. `None` is not an error: it is what tells the front end to take a
    /// fresh snapshot instead of drawing a conversation with a hole in it.
    pub fn since(&self, seq: u64) -> Option<Vec<Event>> {
        // Everything above `trimmed_through` is still here, so a caller holding
        // exactly that number — or anything later — can be caught up, and the
        // newest `seq` is always one of those. Below it something is missing,
        // whether or not the journal's first event happens to be older still.
        //
        // A `seq` past the tip lands here as `Some(vec![])` rather than `None`,
        // and that is deliberate: a journal lives exactly as long as the
        // session id it belongs to, so no client can be holding a cursor minted
        // by another one, and the only way to be ahead of the tip is to have
        // asked twice with nothing appended in between.
        if seq < self.trimmed_through {
            return None;
        }
        Some(self.events.iter().filter(|event| event.seq > seq).cloned().collect())
    }

    pub fn snapshot(&self) -> (Vec<Event>, u64) {
        (self.events.clone(), self.next - 1)
    }

    pub fn events(&self) -> &[Event] {
        &self.events
    }

    /// Drop from the front, and never a question nobody has answered.
    ///
    /// `state_of` folds over whatever the journal holds, so trimming an open
    /// `Permission` away would leave the session `needs-you` with nothing on
    /// screen to answer and no way back to `running`. An answered one carries
    /// no such weight and goes with the rest.
    fn trim(&mut self) {
        if self.events.len() <= BUDGET {
            return;
        }
        let answered: HashSet<String> = self
            .events
            .iter()
            .filter_map(|event| match &event.kind {
                EventKind::PermissionAnswered { id, .. } => Some(id.clone()),
                _ => None,
            })
            .collect();
        let mut over = self.events.len() - BUDGET;
        let mut trimmed_through = self.trimmed_through;
        self.events.retain(|event| {
            if over == 0 {
                return true;
            }
            if let EventKind::Permission { id, .. } = &event.kind {
                if !answered.contains(id) {
                    return true;
                }
            }
            over -= 1;
            trimmed_through = trimmed_through.max(event.seq);
            false
        });
        self.trimmed_through = trimmed_through;
    }
}

impl Default for Journal {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::session::model::{Actor, Decision};

    fn text(body: &str) -> EventKind {
        EventKind::Text { text: body.into() }
    }

    fn delta(piece: &str) -> EventKind {
        EventKind::TextDelta { text: piece.into() }
    }

    fn at() -> String {
        "2026-09-10T12:00:00Z".into()
    }

    fn permission(id: &str) -> EventKind {
        EventKind::Permission {
            id: id.into(),
            tool: "Bash".into(),
            detail: "ls".into(),
            options: vec![],
        }
    }

    #[test]
    fn seq_starts_at_one_so_that_zero_can_mean_nothing_seen_yet() {
        // A front end that has never attached asks from 0, and must be given
        // everything rather than everything-but-the-first.
        let mut journal = Journal::new();
        assert_eq!(journal.append(text("a"), at()).seq, 1);
        assert_eq!(journal.since(0).expect("nothing was trimmed").len(), 1);
    }

    #[test]
    fn since_hands_back_only_what_came_after_the_number_it_was_given() {
        let mut journal = Journal::new();
        journal.append(text("a"), at());
        journal.append(text("b"), at());
        journal.append(text("c"), at());
        let tail = journal.since(1).expect("nothing was trimmed");
        assert_eq!(tail.len(), 2);
        assert_eq!(tail[0].seq, 2);
    }

    #[test]
    fn a_seq_that_was_trimmed_away_answers_none_rather_than_a_hole() {
        // None is what makes the front end re-attach. Handing back a partial
        // tail would leave it drawing a conversation with a silent gap in it.
        let mut journal = Journal::new();
        for n in 0..BUDGET + 10 {
            journal.append(text(&n.to_string()), at());
        }
        assert!(journal.since(1).is_none(), "seq 1 is long gone");
        assert!(journal.since(journal.snapshot().1).is_some(), "the newest is always reachable");
    }

    #[test]
    fn a_pinned_question_does_not_blind_since_to_everything_trimmed_behind_it() {
        // The two safety rules meet here, and apart they each look right: the
        // question is kept at the front, everything after it goes, and the
        // journal stops being a contiguous suffix. A floor read off the oldest
        // event still held is then that pinned `seq` — 1 — and no cursor on
        // earth is below it, so `since` would answer `Some` over a hole four
        // thousand events wide and the conversation would lose them silently.
        let mut journal = Journal::new();
        journal.append(permission("p1"), at());
        for n in 0..BUDGET * 2 {
            journal.append(text(&n.to_string()), at());
        }
        assert!(
            journal.since(3000).is_none(),
            "a cursor inside the gap is told to re-snapshot, not handed the far side of it"
        );
        assert!(
            journal.since(journal.snapshot().1).is_some(),
            "the newest is still always reachable"
        );
    }

    #[test]
    fn a_cursor_at_the_far_end_of_the_number_line_is_answered_rather_than_overflowing() {
        // The floor is compared against, never added to. An implementation
        // asking whether `seq + 1` is below it panics here in a debug build and
        // wraps to zero in a release one.
        let mut journal = Journal::new();
        journal.append(text("a"), at());
        assert_eq!(journal.since(u64::MAX).expect("nothing was trimmed").len(), 0);
    }

    #[test]
    fn trimming_never_takes_a_question_still_waiting_for_an_answer() {
        // Losing it would leave the session stuck in needs-you with nothing on
        // screen to answer, since `state_of` folds over what the journal holds.
        let mut journal = Journal::new();
        journal.append(permission("p1"), at());
        for n in 0..BUDGET + 10 {
            journal.append(text(&n.to_string()), at());
        }
        assert!(
            journal
                .events()
                .iter()
                .any(|e| matches!(&e.kind, EventKind::Permission { id, .. } if id == "p1")),
            "the unanswered question survived the trim"
        );
    }

    #[test]
    fn a_question_that_has_been_answered_is_ordinary_and_may_be_trimmed() {
        let mut journal = Journal::new();
        journal.append(permission("p1"), at());
        journal.append(
            EventKind::PermissionAnswered { id: "p1".into(), decision: Decision::Allow },
            at(),
        );
        for n in 0..BUDGET + 10 {
            journal.append(text(&n.to_string()), at());
        }
        assert!(journal.events().len() <= BUDGET + 2, "answered questions do not accumulate");
    }

    #[test]
    fn the_snapshot_says_which_seq_to_continue_from() {
        let mut journal = Journal::new();
        journal.append(text("a"), at());
        journal.append(text("b"), at());
        let (events, seq) = journal.snapshot();
        assert_eq!(seq, 2);
        assert_eq!(events.last().expect("two were appended").seq, 2);
        assert!(journal.since(seq).expect("nothing trimmed").is_empty());
    }

    #[test]
    fn an_empty_journal_snapshots_to_nothing_at_zero() {
        let (events, seq) = Journal::new().snapshot();
        assert!(events.is_empty());
        assert_eq!(seq, 0);
    }

    #[test]
    fn the_turn_start_of_an_agent_is_kept_the_way_a_person_message_is() {
        // Both are structure rather than chatter, and a journal trimmed to
        // paragraphs alone would draw prose with no turn boundaries in it.
        let mut journal = Journal::new();
        journal.append(EventKind::TurnStart { by: Actor::Person }, at());
        assert_eq!(journal.events().len(), 1);
    }

    /// The blocking finding smetana-6we6's review came back with: a
    /// substantial streamed reply is several hundred `TextDelta`s where it
    /// used to be one `Text`, and `BUDGET` would arrive in five to fifteen
    /// turns rather than the week `BUDGET`'s own comment promises. The closing
    /// `Text` has to take its deltas with it.
    #[test]
    fn a_closing_text_drops_the_deltas_it_superseded() {
        let mut journal = Journal::new();
        journal.append(EventKind::TurnStart { by: Actor::Agent }, at());
        journal.append(delta("The identity"), at());
        journal.append(delta(" is read once"), at());
        journal.append(text("The identity is read once."), at());

        let kinds: Vec<&EventKind> = journal.events().iter().map(|event| &event.kind).collect();
        assert_eq!(
            kinds,
            vec![
                &EventKind::TurnStart { by: Actor::Agent },
                &EventKind::Text { text: "The identity is read once.".into() }
            ],
            "the deltas must not go on living beside the reply that closed over them"
        );
    }

    /// A reply that streamed for a long time can still be several hundred
    /// deltas wide the instant before it closes — this is the shape that
    /// blew the budget open, pinned directly rather than only through its
    /// consequence above.
    #[test]
    fn many_deltas_collapse_to_one_event_on_close() {
        let mut journal = Journal::new();
        for n in 0..500 {
            journal.append(delta(&n.to_string()), at());
        }
        assert_eq!(journal.events().len(), 500);
        journal.append(text("the whole reply"), at());
        assert_eq!(journal.events().len(), 1);
    }

    /// Only the run just superseded goes — an earlier, already-closed reply's
    /// words stay exactly as they were, and a *second* stream's deltas are
    /// left standing until their own close, not swept early by someone else's.
    #[test]
    fn closing_one_stream_never_touches_an_earlier_reply_or_a_later_one() {
        let mut journal = Journal::new();
        journal.append(text("the first reply"), at());
        journal.append(delta("the second"), at());
        journal.append(text("the second reply"), at());
        journal.append(delta("a third, still open"), at());

        let kinds: Vec<&EventKind> = journal.events().iter().map(|event| &event.kind).collect();
        assert_eq!(
            kinds,
            vec![
                &EventKind::Text { text: "the first reply".into() },
                &EventKind::Text { text: "the second reply".into() },
                &EventKind::TextDelta { text: "a third, still open".into() }
            ]
        );
    }

    /// `trimmed_through` is deliberately untouched by the collapse: it is
    /// `since`'s floor, and moving it would tell a cursor holder to
    /// re-snapshot over a gap that cost it nothing, since it already had the
    /// deltas live before they were dropped.
    #[test]
    fn dropping_superseded_deltas_does_not_move_the_since_floor() {
        let mut journal = Journal::new();
        journal.append(delta("a"), at());
        journal.append(delta("b"), at());
        let (_, before_close) = journal.snapshot();
        journal.append(text("ab"), at());

        assert!(
            journal.since(before_close).is_some(),
            "a cursor that watched the stream live is not told to re-snapshot"
        );
        assert_eq!(journal.since(before_close).expect("checked above").len(), 1);
    }

    /// A fresh attacher after the close sees the consolidated reply and none
    /// of the deltas it was built from — the live-journal half of the
    /// acceptance criterion the transcript-replay half already covered.
    #[test]
    fn a_snapshot_taken_after_the_close_holds_no_deltas_at_all() {
        let mut journal = Journal::new();
        journal.append(EventKind::TurnStart { by: Actor::Agent }, at());
        for piece in ["The ", "identity ", "is read once."] {
            journal.append(delta(piece), at());
        }
        journal.append(text("The identity is read once."), at());

        let (events, _) = journal.snapshot();
        assert!(
            !events.iter().any(|event| matches!(event.kind, EventKind::TextDelta { .. })),
            "a fresh attacher must never see a delta the reply it belongs to has already closed"
        );
    }
}

